packages/core/src/execution-engine/ssh-clients-manager.ts

```diff
@@ -11,21 +12,54 @@ class SSHClientsConfig {
 	idleTimeout: number = 5 * 60;
 }
 
+type Registration = {
+	client: Client;
+
+	/**
+	 * We keep this timestamp to check if a client hasn't been used in a while,
+	 * and if it needs to be closed.
+	 */
+	lastUsed: Date;
+
+	abortController: AbortController;
+};
+
 @Service()
 export class SSHClientsManager {
-	readonly clients = new Map<string, { client: Client; lastUsed: Date }>();
+	readonly clients = new Map<string, Registration>();
+
+	readonly clientsReversed = new WeakMap<Client, string>();
 
-	constructor(private readonly config: SSHClientsConfig) {
+	private cleanupTimer: NodeJS.Timeout;
+
+	constructor(
+		private readonly config: SSHClientsConfig,
+		private readonly logger: Logger,
+	) {
 		// Close all SSH connections when the process exits
 		process.on('exit', () => this.onShutdown());
 
-		if (process.env.NODE_ENV === 'test') return;
-
 		// Regularly close stale SSH connections
-		setInterval(() => this.cleanupStaleConnections(), 60 * 1000);
+		this.cleanupTimer = setInterval(() => this.cleanupStaleConnections(), 60 * 1000);
+
+		this.logger = logger.scoped('ssh-client');
+	}
+
+	updateLastUsed(client: Client) {
+		const key = this.clientsReversed.get(client);
```

Comment:
We're silently ignoring an unregistered client, should we log this? Does it hide a logic error elsewhere to call this method with an unregistered client?

Reply:
> We're silently ignoring an unregistered client, should we log this?

We could log it. I'll make a follow up for this.

> Does it hide a logic error elsewhere to call this method with an unregistered client?

Probably not. As long as `SSHClientsManager.cleanupStaleConnections` and `ConnectionPoolManager.cleanupStaleConnections` are sync there shouldn't be a problem, like having cleaned up clients in the map or having still connected clients in not in the map.

Reply:
done: e4c0fa967e619ea7cd629269bfcfa5af55012aab